### PR TITLE
fix(insights): detect clicks on children of [data-insights] HTML elements

### PR DESCRIPTION
**Pull request by tkrugg**
**Summary**

When the insights integration is used in the following way:

```js
<a href="#" ${insights('clickObjectIDsAfterSearch', {...})}>
   <h1> ${hit.name} </h1>
</a>
```
and click occurs on `h1`, insight client is not called.

This PR addresses this problem.

### CODE REVIEW

File: src/lib/insights/listener.tsx [hunk 1]
```diff
@@ -9,23 +9,37 @@ type WithInsightsListenerProps = {
...
-
       if (!props.insights) {
         throw new Error(
           'The `insightsClient` option has not been provided to `instantsearch`.'
         );
       }
-
-      const { method, payload } = readDataAttributes(
-        event.target as HTMLElement
+      const realTarget = findRealTarget(
```
Comment (tkrugg) [inline src/lib/insights/listener.tsx]: A better naming for this variable is welcome 😄

File: src/lib/__tests__/insights-listener-test.tsx [hunk 2]
```diff
@@ -61,6 +61,62 @@ describe('withInsightsListener', () => {
...
+    const HitsWithInsightsListener: any = withInsightsListener(Hits);
+    const { container } = render(
+      <HitsWithInsightsListener
+        hits={hits}
+        results={results}
+        instantSearchInstance={instantSearchInstance}
+        insights={insights}
+      />
+    );
+    const innerChild = container.querySelector('h1') as HTMLElement;
```
Comment (francoischalifour) [inline src/lib/__tests__/insights-listener-test.tsx]:
```suggestion
    const innerChild = container.querySelector<HTMLElement>('h1');
```

File: src/lib/insights/listener.tsx (hunk 1 above)
Comment (francoischalifour) [reply src/lib/insights/listener.tsx]: `insightsRoot` seems fine.

File: src/lib/insights/listener.tsx [hunk 3]
```diff
@@ -9,23 +9,37 @@ type WithInsightsListenerProps = {
   insights: InsightsClientWrapper;
 };
 
+const findRealTarget = (
```
Comment (francoischalifour) [inline src/lib/insights/listener.tsx]: We could name this `findInsightsRoot`.

File: src/lib/insights/listener.tsx [hunk 4]
```diff
@@ -9,23 +9,37 @@ type WithInsightsListenerProps = {
...
+  startElement: HTMLElement | null,
+  endElement: HTMLElement | null
+): HTMLElement | null => {
+  let element: HTMLElement | null = startElement;
```
Comment (francoischalifour) [inline src/lib/insights/listener.tsx]:
```suggestion
  let element = startElement;
```

File: src/lib/insights/listener.tsx (hunk 1 above)
Comment (tkrugg) [reply src/lib/insights/listener.tsx]: what would `root` mean here?

Comment (francoischalifour) [reply src/lib/insights/listener.tsx]: What `findRealTarget` (current naming) is doing here is resolving the root HTML element where data attributes are defined. The root is where the data needs to be read from.

`insightsRootTarget` if you want to keep the "target" name.

Comment (tkrugg) [reply src/lib/insights/listener.tsx]: Ok I understand, but actually the code we're adding here is trying to find a DOM elements (not necessarily a root element) that carries the data-insights attribute, so I'm not sure how `root` adds meaning here.
Is `insightsTarget` fine or would it be something missing?

Comment (francoischalifour) [reply src/lib/insights/listener.tsx]: They're both fine. Root just means it's the highest parent in this context.

Comment (tkrugg) [reply src/lib/insights/listener.tsx]: ok, renamed to `insightsTarget` in 115c932f

File: src/lib/insights/listener.tsx (hunk 3 above)
Comment (tkrugg) [reply src/lib/insights/listener.tsx]: renamed to `findInsightsTarget` in 115c932f as agreed at https://github.com/algolia/instantsearch.js/pull/4197#discussion_r351394715

**Pull request merged**